server/src/main/java/org/opensearch/cluster/service/ClusterManagerTaskThrottler.java

```diff
@@ -0,0 +1,200 @@
+    private final ConcurrentMap<String, Long> tasksThreshold;
+    private final Supplier<Version> minNodeVersionSupplier;
+
+    public ClusterManagerTaskThrottler(
+        final ClusterSettings clusterSettings,
+        final Supplier<Version> minNodeVersionSupplier,
+        final ClusterManagerTaskThrottlerListener clusterManagerTaskThrottlerListener
+    ) {
+        clusterSettings.addSettingsUpdateConsumer(THRESHOLD_SETTINGS, this::updateSetting, this::validateSetting);
+        this.minNodeVersionSupplier = minNodeVersionSupplier;
+        this.clusterManagerTaskThrottlerListener = clusterManagerTaskThrottlerListener;
+        tasksCount = new ConcurrentHashMap<>(128); // setting initial capacity so each task will land in different segment
+        tasksThreshold = new ConcurrentHashMap<>(128); // setting initial capacity so each task will land in different segment
+    }
+
+    /**
+     * To configure a new task for throttling,
+     * * Register task to cluster service with task key,
+     * * override getClusterManagerThrottlingKey method with above task key in task executor.
+     * * Verify that throttled tasks would be retried from data nodes
+     *
+     * Added retry mechanism in TransportClusterManagerNodeAction, so it would be retried for customer generated tasks.
+     *
+     * If tasks are not getting retried then we can register with false flag, so user won't be able to configure threshold limits for it.
+     */
+    protected ThrottlingKey registerClusterManagerTask(String taskKey, boolean throttlingEnabled) {
+        ThrottlingKey throttlingKey = new ThrottlingKey(taskKey, throttlingEnabled);
+        if (THROTTLING_TASK_KEYS.containsKey(taskKey)) {
+            throw new IllegalArgumentException("There is already a Throttling key registered with same name: " + taskKey);
+        }
+        THROTTLING_TASK_KEYS.put(taskKey, throttlingKey);
+        return throttlingKey;
+    }
+
+    /**
+     * Class to store the throttling key for the tasks of cluster manager
+     */
+    public static class ThrottlingKey {
+        private String taskThrottlingKey;
+        private boolean throttlingEnabled;
+
+        /**
+         * Class for throttling key of tasks
+         *
+         * @param taskThrottlingKey - throttling key for task
+         * @param throttlingEnabled - if throttling is enabled or not i.e. data node is performing retry over throttling exception or not.
+         */
+        private ThrottlingKey(String taskThrottlingKey, boolean throttlingEnabled) {
+            this.taskThrottlingKey = taskThrottlingKey;
+            this.throttlingEnabled = throttlingEnabled;
+        }
+
+        public String getTaskThrottlingKey() {
+            return taskThrottlingKey;
+        }
+
+        public boolean isThrottlingEnabled() {
+            return throttlingEnabled;
+        }
+    }
+
+    void validateSetting(final Settings settings) {
+        if (minNodeVersionSupplier.get().compareTo(Version.V_2_4_0) < 0) {
```

Comment:
We might need immediate backport to avoid BWC failures

Reply:
We need to back port it to 2.x branch only right? Will raise PR for back porting to 2.x branch.